plugins/mongodb/mongodbembed.cpp

```diff
@@ -0,0 +1,2287 @@
+
+        MongoDBConnection::instance().configure(std::move(instance), bsoncxx::stdx::make_unique<mongocxx::pool>(std::move(uri)));
+    }
+
+    /**
+     * @brief Construct a new MongoDBEmbedFunctionContext object
+     * 
+     * @param _logctx Context logger for use with the MongoDBRecordBinder
+     * MongoDBDatasetBinder classes.
+     * @param options Pointer to the list of options that are passed into the Embed function.
+     * This contains the server name, username, password, etc.
+     * @param _flags Should be zero if the embedded script is ok.
+     */
+    MongoDBEmbedFunctionContext::MongoDBEmbedFunctionContext(const IContextLogger &_logctx, const char *options, unsigned _flags)
+    : logctx(_logctx), m_NextRow(), m_nextParam(0), m_numParams(0), m_scriptFlags(_flags)
+    {
+        const char *server = "";
+        const char *user = "";
+        const char *password = "";
+        const char *databaseName = "";
+        const char *collectionName = "";
+        const char *connectionOptions = "";
+
+        unsigned port = 0;
+        unsigned batchSize = 0;
+        bool useSSL = false;
+        StringBuffer connectionString;
+
+        StringArray inputOptions;
+        inputOptions.appendList(options, ",");
+        ForEachItemIn(idx, inputOptions) 
+        {
+            const char *opt = inputOptions.item(idx);
+            const char *val = strchr(opt, '=');
+            if (val)
+            {
+                StringBuffer optName(val-opt, opt);
+                val++;
+                if (stricmp(optName, "server")==0)
+                    server = val;
+                else if (stricmp(optName, "port")==0)
+                    port = atoi(val);
+                else if (stricmp(optName, "user")==0)
+                    user = val;
+                else if (stricmp(optName, "password")==0)
+                    password = val;
+                else if (stricmp(optName, "database")==0)
+                    databaseName = val;
+                else if (stricmp(optName, "collection")==0)
+                    collectionName = val;
+                else if (stricmp(optName, "batchSize")==0)
+                    batchSize = atoi(val);
+                else if (stricmp(optName, "connectionOptions") ==0)
+                    connectionOptions = val;
+                else
+                    failx("Unknown option %s", optName.str());
+            }
+        }
+
+        if (!isEmptyString(server) && port !=0)
+            failx("Specifying a port is not allowed when connecting to an Atlas Cluster.");
+        else if (port != 0)
+            connectionString.appendf("mongodb://localhost:%i/%s", port, connectionOptions); // Connection to a local MongoDB instance
+        else
+            connectionString.appendf("mongodb+srv://%s:%s@%s/?retryWrites=true&w=majority%s", user, password, server, connectionOptions); // Connection to MongoDB Atlas server
```

Comment:
I think it's possible for user and pass to be empty, we should either require those values, or avoid the user:pass@ portion in that scenario.

Reply:
Correct they can be empty when connecting to a local instance, but a username and password are required to connect to a MongoDB cluster. I will change it so that they are required for connecting remotely and not allowed when connecting locally.